src/Compilers/VisualBasic/Portable/BoundTree/BoundFieldAccess.vb

```diff
@@ -33,6 +33,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
 
         Public Overrides ReadOnly Property ConstantValueOpt As ConstantValue
             Get
+                ' decimal and datetime const fields require one-time synthesized assignment in cctor
+                ' when used as a LHS of an assignment, the access is not a const.
+                If _IsLValue Then
+                    Return Nothing
```

Comment:
The field does not need to be decimal or datetime. Any field access that is an lvalue should not have constant value. It just so happens that only decimal and datetime fields can be formally constants while also participating in assignments. An assert for such condition should not be here, but somewhere at node construction. 
I think it is not worth it though, since it would basically guard that we are not assigning to constants in general, which seems improbable for many other reasons.